### ISSUE TITLE
HTML <table> like column width (columns are as wide as the columns widest cell content)

### ISSUE THREAD

**Issue by emattias** [bvaughn/react-virtualized#468]
I think it would be interesting to create react-virtualized tables where the columns adjust to the content, like regular HTML tables. 

Not sure if this can be accomplished using only flexbox. [Here's a JSFiddle](https://jsfiddle.net/tacc2tqr/1/) with an interesting solution.

I have a use case where I have one column with varying content width and 3 that have a set max-width.

**Comment by bvaughn** [comment]
Hey @emattias,

I agree this would be a useful component. :smile:

Unfortunately there is a key problem due to the fact that we are windowing data (and so the table only ever contains a slide of the total rows). As a user scrolls the table and columns get new data- they would also resize- which is a pretty funky user experience. 😅

Unfortunately the only way to really achieve this would be to pre-measure every column which would negate the performance benefits of windowing.

It's something I've briefly thought about in the past but see no performant way to achieve, so...I'm going to close this issue for now. Hope you understand! Happy to listen to ideas or suggestions if you have any that I haven't thought of though. 😁

**Comment by emattias** [comment]
Thanks @bvaughn 

My use case is a scrollable data table that when you scroll update a D3 graph (using [Victory.js](https://formidable.com/open-source/victory/)).

When I use react-virtualized for the table it is much less laggy (using onScroll on the Table component) than using a regular HTML Table. I tried switching to HTML table because I didn't want the content cut off.

Do you have an idea of how react-virtualized avoids lag? I have a couple ideas:

- Are you using requestanimationframe and/or some other debounce solution in the scroll-listener?
- Are you avoiding layout thrashing when getting the scroll-position? [Since scrollTop for example causes layout reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a)
- Is it just the fact that the whole table is absolute positioned? Does the HTML table do some layout calculations when scrolling?
- Something else that you might know? I am very interested in anything that you might know about this!

Either way this really shows the awesome performance work that is done by react-virtualized!

**Comment by bvaughn** [comment]
The lag you're referring to is probably due to having too many DOM elements in your table- particularly if there is also a complicated graph involved. react-virtualized is fast because it only renders what's visible. (See [slides from my recent presentation](https://bvaughn.github.io/connect-tech-2016/#/8/1) for a visual explanation if that's helpful.)

I am also using `requestAnimationFrame` (which itself debounces scrolling events slightly), as well as dedicated layers for each cell to avoid repainting the whole table on scroll. Lots of small optimizations go into making a lib like this fast. 😄

Thank you!

**Comment by emattias** [comment]
Thanks for the links to the slides! Amazing work!👌

I discovered that I have considerable lag with react-virtualized Table in firefox.. That work is not prioritized right now but I will investigate later!

**Comment by steviesama** [comment]
@bvaughn I'm currently using fixed-data-table. I don't know if I'm gonna hit the wall with this thing or not, but some of my columns are much narrower than others. Is it possible to just virtualize y, and allow the specifying fixed varying column widths as with fixed-data-table?

**Comment by bvaughn** [comment]
I'm not sure I understand what you're asking. If it's the same thing as was originally requested, then [my initial response](https://github.com/bvaughn/react-virtualized/issues/468#issuecomment-260970405) would still apply. If it's something else, could you maybe reword or give an example to help me understand the difference?

**Comment by steviesama** [comment]
@bvaughn Actually it wasn't precisely what was originally requested from my understanding, but aside from that were was a misunderstanding on my part.  I hadn't really taken a good look at all the controls you had in the repo. I don't recall the exact control I was looking at, but I think it was a grid where both x and y were virtualized, and I was thinking about a table with a limited number of columns with varying widths where only the y axis was virtualized. When spent some more time looking, your table control actually does exactly what I want with column widths. Eventually I'd like to switch over from fixed-data-table2 to it. Is there any column resize functionality for the table control?

**Comment by bvaughn** [comment]
> Is there any column resize functionality for the table control?

Not built in, no. That's not really core to windowing so it's not part of the lib.

You could use a library like `react-draggable` and an approach like [this](https://bvaughn.github.io/forward-js-2017/#/28/1) though (source [here](https://github.com/bvaughn/forward-js-2017/blob/master/src/Components/ResizableRowsList.js)).
